### PR TITLE
Eliminate unnecessary GC handle allocations in Unix Sockets implementation

### PR DESCRIPTION
**Pull request by ericeil**
Replaces GC handle allocation with a pinned local, in the cases where we are pinning a single buffer over a statically-scoped region.  This should have significantly lower runtime overhead.

@stephentoub, @pgavlin

### CODE REVIEW

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs [hunk 1]
```diff
@@ -384,9 +343,9 @@ private static unsafe int Receive(int fd, SocketFlags flags, int available, ILis
             var handles = new GCHandle[maxBuffers];
             var iovecs = new Interop.Sys.IOVector[maxBuffers];
 
-            var pinnedSocketAddress = default(GCHandle);
-            byte* sockAddr = null;
             int sockAddrLen = 0;
+            if (socketAddress != null)
```
Comment (pgavlin) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs]: Please use braces around the body of this if statement.

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs [hunk 2]
```diff
@@ -289,9 +259,9 @@ private static unsafe int Send(int fd, SocketFlags flags, IList<ArraySegment<byt
             // Pin buffers and set up iovecs.
             int startIndex = bufferIndex, startOffset = offset;
 
-            var pinnedSocketAddress = default(GCHandle);
-            byte* sockAddr = null;
             int sockAddrLen = 0;
+            if (socketAddress != null)
```
Comment (pgavlin) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs]: Same here.

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs [hunk 3]
```diff
@@ -231,46 +216,31 @@ private static unsafe int Receive(int fd, SocketFlags flags, int available, byte
 
         private static unsafe int Send(int fd, SocketFlags flags, byte[] buffer, ref int offset, ref int count, byte[] socketAddress, int socketAddressLen, out Interop.Error errno)
         {
-            var pinnedSocketAddress = default(GCHandle);
-            byte* sockAddr = null;
+            int sent;
+
             int sockAddrLen = 0;
+            if (socketAddress != null)
```
Comment (pgavlin) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs]: And here.

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs [hunk 4]
```diff
@@ -179,45 +179,30 @@ private static unsafe int Receive(int fd, SocketFlags flags, int available, byte
         {
             Debug.Assert(socketAddress != null || socketAddressLen == 0);
 
-            var pinnedSocketAddress = default(GCHandle);
-            byte* sockAddr = null;
+            long received;
+
             int sockAddrLen = 0;
+            if (socketAddress != null)
```
Comment (pgavlin) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs]: ...and here. :)

**Comment by pgavlin** [comment]
Good catch. LGTM aside from the formatting issues.

**Comment by stephentoub** [comment]
LGTM

**Pull request merged**